### PR TITLE
PWN-7020 - jupiter tokens request

### PR DESCRIPTION
**Pull request by NikolaiEmelianov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7020

## Description of Work

jupiter tokens request

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/swap/jupiter/repository/model/JupiterToken.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package org.p2p.wallet.swap.jupiter.repository.model
+
+import org.p2p.wallet.swap.jupiter.api.response.tokens.JupiterTokenExtensions
+
+data class JupiterToken(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/repository/model/JupiterToken.kt:5]: we can remove implementation detail - Jupiter and call it just SwapToken? 
just a nitpick

File: app/src/main/java/org/p2p/wallet/swap/jupiter/repository/model/JupiterToken.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    val address: String,
+    val chainId: Int,
+    val decimals: Int,
+    val extensions: JupiterTokenExtensions,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/repository/model/JupiterToken.kt:9]: domain models can't have entities with @SerializedName inside

File: app/src/main/java/org/p2p/wallet/swap/jupiter/repository/model/JupiterToken.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package org.p2p.wallet.swap.jupiter.repository.model
+
+import org.p2p.wallet.swap.jupiter.api.response.tokens.JupiterTokenExtensions
+
+data class JupiterToken(
+    val address: String,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/repository/model/JupiterToken.kt:6]: if it's mint - it's Base58String

File: app/src/main/java/org/p2p/wallet/swap/jupiter/repository/model/JupiterToken.kt [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    val chainId: Int,
+    val decimals: Int,
+    val extensions: JupiterTokenExtensions,
+    val logoURI: String,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/repository/model/JupiterToken.kt:10]: logoUri

**Review by gslevinkov — APPROVED** [review]

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/swap/jupiter/repository/JupiterSwapTokenRepository.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    private val api: SwapJupiterTokensApi,
+    private val dispatchers: CoroutineDispatchers,
+) : SwapTokensRepository {
+
+    override suspend fun getTokens(): List<SwapToken> =
+        withContext(dispatchers.io) {
+            api.getTokens().toJupiterToken()
+        }
+
+    private fun List<JupiterTokenResponse>.toJupiterToken(): List<SwapToken> = map { respons ->
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/repository/JupiterSwapTokenRepository.kt:21]: move mappings to a separate mapper

File: app/src/main/java/org/p2p/wallet/swap/jupiter/repository/JupiterSwapTokenRepository.kt [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import org.p2p.wallet.utils.toBase58Instance
+import kotlinx.coroutines.withContext
+
+class JupiterSwapTokenRepository(
+    private val api: SwapJupiterTokensApi,
+    private val dispatchers: CoroutineDispatchers,
+) : SwapTokensRepository {
+
+    override suspend fun getTokens(): List<SwapToken> =
+        withContext(dispatchers.io) {
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/repository/JupiterSwapTokenRepository.kt:17]: this context is redundant

File: app/src/main/java/org/p2p/wallet/swap/jupiter/repository/model/SwapToken.kt [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+data class SwapToken(
+    val address: Base58String,
+    val chainId: Int,
+    val decimals: Int,
+    val extensions: SwapTokenExtensions,
+    val logoUri: String,
+    val name: String,
+    val symbol: String,
+    val tags: List<String>
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/repository/model/SwapToken.kt:13]: Are you sure we don't have `SwapToken` anymore in the project?
Seems to me that this is a conflict name. 

Let's name it `JupiterSwapToken` please

File: app/src/main/java/org/p2p/wallet/swap/jupiter/repository/model/SwapTokenExtensions.kt [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package org.p2p.wallet.swap.jupiter.repository.model
+
+data class SwapTokenExtensions(
+    val coingeckoId: String
+)
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/repository/model/SwapTokenExtensions.kt:5]: This class is redundant.
Use `coingeckoId` directly in `SwapToken`

**Review by 1sakov — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/p2p/wallet/swap/jupiter/repository/JupiterSwapTokenRepository.kt [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+import org.p2p.wallet.infrastructure.dispatchers.CoroutineDispatchers
+import org.p2p.wallet.swap.jupiter.api.SwapJupiterTokensApi
+import org.p2p.wallet.swap.jupiter.api.response.tokens.JupiterTokenResponse
+import org.p2p.wallet.swap.jupiter.repository.model.JupiterToken
+import kotlinx.coroutines.withContext
+
+class JupiterSwapTokenRepository(
+    private val api: SwapJupiterTokensApi,
+    private val dispatchers: CoroutineDispatchers,
+) : SwapTokensRepository {
```
[the remark this answers is not in the document]
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/repository/JupiterSwapTokenRepository.kt:14]: Why? This repository is getting us SwapTokens - that's all we need
The place where we getting this tokens - is implementation detail, look at the TokenPricesRepository, seems great

[the remark this answers is not in the document]
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/repository/JupiterSwapTokenRepository.kt:14]: The implementation of each swap is too different. 
You will not be able to make as in TokenPrices. 

Orca is absolutely different, Serum swap is different. 
Jupiter has its own implementation.

It's better to know what exactly repository and swap you are using. 
I don't think this pattern of naming is correct here. 
Let's name it with full context.

Mb `JupiterTokensRepository` is better

**Review by 1sakov — APPROVED** [review]

**Pull request merged**